vta/hardware/chisel/src/main/scala/Core.scala

```diff
@@ -0,0 +1,161 @@
+package vta
+
+import chisel3._
+import freechips.rocketchip.config.{Parameters, Field}
+
+case object LOG_INP_WIDTH extends Field[Int]
+case object LOG_WGT_WIDTH extends Field[Int]
+case object LOG_ACC_WIDTH extends Field[Int]
+case object LOG_OUT_WIDTH extends Field[Int]
+case object LOG_BATCH extends Field[Int]
+case object LOG_BLOCK_IN extends Field[Int]
+case object LOG_BLOCK_OUT extends Field[Int]
+case object LOG_UOP_BUFF_SIZE extends Field[Int]
+case object LOG_INP_BUFF_SIZE extends Field[Int]
+case object LOG_WGT_BUFF_SIZE extends Field[Int]
+case object LOG_ACC_BUFF_SIZE extends Field[Int]
+
+class AvalonSlaveIO(val dataBits: Int = 32, val addrBits: Int = 1)(implicit p: Parameters) extends CoreBundle()(p) {
+  val waitrequest = Output(Bool())
+  val address = Input(UInt(addrBits.W))
+  val read = Input(Bool())
+  val readdata = Output(UInt(dataBits.W))
+  val write = Input(Bool())
+  val writedata = Input(UInt(dataBits.W))
+}
+  
+class AvalonSinkIO(val dataBits: Int = 32)(implicit p: Parameters) extends CoreBundle()(p) {
+  val ready = Output(Bool())
+  val valid = Input(Bool())
+  val data = Input(UInt(dataBits.W))
+}
+
+abstract trait CoreParams {
+  implicit val p: Parameters
+
+  val log_uop_buff_size = p(LOG_UOP_BUFF_SIZE)
+  val log_inp_buff_size = p(LOG_INP_BUFF_SIZE)
+  val log_wgt_buff_size = p(LOG_WGT_BUFF_SIZE)
+  val log_acc_buff_size = p(LOG_ACC_BUFF_SIZE)
+
+  val log_batch = p(LOG_BATCH)
+  val log_block_in = p(LOG_BLOCK_IN)
+  val log_block_out = p(LOG_BLOCK_OUT)
+  val batch = 1 << p(LOG_BATCH)
+  val block_in = 1 << p(LOG_BLOCK_IN)
+  val block_out = 1 << p(LOG_BLOCK_OUT)
+
+  val log_uop_width = 5
+  val log_inp_width = p(LOG_INP_WIDTH)
+  val log_wgt_width = p(LOG_WGT_WIDTH)
+  val log_acc_width = p(LOG_ACC_WIDTH)
+  val log_out_width = p(LOG_OUT_WIDTH)
+  val uop_width = 1 << log_uop_width
+  val inp_width = 1 << p(LOG_INP_WIDTH)
+  val wgt_width = 1 << p(LOG_WGT_WIDTH)
+  val acc_width = 1 << p(LOG_ACC_WIDTH)
+  val out_width = 1 << p(LOG_OUT_WIDTH)
+
+  val opcode_bit_width = 3
+  val alu_opcode_bit_width = 2
+  val memop_id_bit_width = 2
+  val memop_sram_addr_bit_width = 16
+  val memop_dram_addr_bit_width = 32
+  val memop_size_bit_width = 16
+  val memop_stride_bit_width = 16
+  val memop_pad_bit_width = 4
+  val memop_pad_val_bit_width = 2
+  val aluop_imm_bit_width = 16
+  val loop_iter_width = 14
+
+  val mem_id_uop = 0
+  val mem_id_wgt = 1
+  val mem_id_inp = 2
+  val mem_id_acc = 3
+  val mem_id_out = 4
+
+  val opcode_load = 0
+  val opcode_store = 1
+  val opcode_gemm = 2
+  val opcode_finish = 3
+  val opcode_alu = 4
```

Comment:
If you're not averse to using the `chisel3.experimental` package, you can replace these ints with strongly typed [`ChiselEnum`](https://github.com/freechipsproject/chisel3/blob/5509cdd4c8332c53151e10ba5bdbe0684af1c05b/src/test/scala/chiselTests/StrongEnum.scala#L12)

Reply:
The intention to hard write the `opcode` values is that I wanted to make current VTA implement completely compatible with the existing one, so that there is no need to revise current software stack. (I'm not quite sure whether `ChiselEnum` would start to enumerate from 1 or zero.)

Reply:
you can start them from either. Generally, it's

```scala
object Opcode extends ChiselEnum {
  val store = Value(1.U)
  val gemm = Value(2.U)
  val finish = Value(3.U)
  val alu = Value(4.U)
}
```

or, more concisely
```scala
object Opcode extends ChiselEnum {
  val _, store, gemm, finish, alu = Value
}
```